Compilation error with MSVC 19.37

From https://github.com/snitch-org/snitch/actions/runs/6781475887/job/18431859675.

The Windows pipeline worked fine last week, with MSVC 19.35.32217.1.
Now, running the CI pipeline with MSVC 19.37.32825.0, we get this: 
```
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): error C7595: 'std::_Literal_zero::_Literal_zero': call to immediate function is not a constant expression [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): message : failure was caused by a read of a variable outside its lifetime [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): message : see usage of 'rhs' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(156,30): message : see reference to function template instantiation 'bool snitch::impl::operator_not_equal::operator ()<T,U>(const T &,const U &) noexcept const' being compiled [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
          with
          [
              T=std::strong_ordering,
              U=int
          ]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(151,26): message : while compiling class template member function 'snitch::impl::expression snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>::to_expression(void) noexcept const' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
D:\a\snitch\snitch\tests\runtime_tests\check.cpp(544,13): message : see the first reference to 'snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>::to_expression' in 'DOCTEST_ANON_FUNC_81' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
D:\a\snitch\snitch\tests\runtime_tests\check.cpp(544,13): message : see reference to class template instantiation 'snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>' being compiled [D:\a\snitch\snitch\build\tests\snitch_runtime_tests.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
  cli.cpp
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): error C7595: 'std::_Literal_zero::_Literal_zero': call to immediate function is not a constant expression [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): message : failure was caused by a read of a variable outside its lifetime [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(36,1): message : see usage of 'rhs' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(156,30): message : see reference to function template instantiation 'bool snitch::impl::operator_not_equal::operator ()<T,U>(const T &,const U &) noexcept const' being compiled [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
          with
          [
              T=std::strong_ordering,
              U=int
          ]
D:\a\snitch\snitch\include\snitch/snitch_expression.hpp(151,26): message : while compiling class template member function 'snitch::impl::expression snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>::to_expression(void) noexcept const' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
D:\a\snitch\snitch\tests\runtime_tests\check.cpp(544,13): message : see the first reference to 'snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>::to_expression' in 'test_fun_34' [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
D:\a\snitch\snitch\tests\runtime_tests\check.cpp(544,13): message : see reference to class template instantiation 'snitch::impl::extracted_binary_expression<true,T,snitch::impl::operator_not_equal,int>' being compiled [D:\a\snitch\snitch\build\tests\snitch_runtime_tests_self.vcxproj]
          with
          [
              T=std::strong_ordering
          ]
```

Minimal reproduction of the issue:
```c++
#include <compare>

struct operator_not_equal {
    template<typename T, typename U>
    constexpr bool operator()(const T& lhs, const U& rhs) const noexcept(noexcept(lhs != rhs))
        requires(requires(const T& lhs, const U& rhs) { lhs != rhs; })
    {
        return lhs != rhs;
    }
};

int main() {
    operator_not_equal{}(1 <=> 1, 0);
    return 0;
}
```

Yields:
```
<source>(9): error C7595: 'std::_Literal_zero::_Literal_zero': call to immediate function is not a constant expression
<source>(9): note: failure was caused by a read of a variable outside its lifetime
<source>(9): note: see usage of 'rhs'
<source>(14): note: see reference to function template instantiation 'bool operator_not_equal::operator ()<std::strong_ordering,int>(const T &,const U &) noexcept const' being compiled
        with
        [
            T=std::strong_ordering,
            U=int
        ]
```

Seems like a compiler bug:
 1. The error message is incorrect; there is no lifetime issue.
 2. If there was really a problem compiling the `!=` expression, the `requires()` should have prevented instantiation of the function. Instead it passed, and the error occurred on instantiation.

--- a/include/snitch/snitch_macros_exceptions.hpp
+++ b/include/snitch/snitch_macros_exceptions.hpp
@@ -10,13 +10,11 @@
 
 #    define SNITCH_REQUIRE_THROWS_AS_IMPL(MAYBE_ABORT, EXPRESSION, ...)                            \
         do {                                                                                       \
-            auto& SNITCH_CURRENT_TEST = snitch::impl::get_current_test();                          \
+            auto& SNITCH_CURRENT_TEST        = snitch::impl::get_current_test();                   \
+            bool  SNITCH_NO_EXCEPTION_THROWN = false;                                              \
             try {                                                                                  \
                 static_cast<void>(EXPRESSION);                                                     \
-                SNITCH_CURRENT_TEST.reg.report_assertion(                                          \
-                    false, SNITCH_CURRENT_TEST, SNITCH_CURRENT_LOCATION,                           \
-                    #__VA_ARGS__ " expected but no exception thrown");                             \
-                MAYBE_ABORT;                                                                       \
+                SNITCH_NO_EXCEPTION_THROWN = true;                                                 \
             } catch (const __VA_ARGS__&) {                                                         \
                 SNITCH_CURRENT_TEST.reg.report_assertion(                                          \
                     true, SNITCH_CURRENT_TEST, SNITCH_CURRENT_LOCATION,                            \
@@ -37,6 +35,12 @@
                     MAYBE_ABORT;                                                                   \
                 }                                                                                  \
             }                                                                                      \
+            if (SNITCH_NO_EXCEPTION_THROWN) {                                                      \
+                SNITCH_CURRENT_TEST.reg.report_assertion(                                          \
+                    false, SNITCH_CURRENT_TEST, SNITCH_CURRENT_LOCATION,                           \
+                    #__VA_ARGS__ " expected but no exception thrown");                             \
+                MAYBE_ABORT;                                                                       \
+            }                                                                                      \
         } while (0)
 
 #    define SNITCH_REQUIRE_THROWS_AS(EXPRESSION, ...)                                              \
@@ -46,13 +50,11 @@
 
 #    define SNITCH_REQUIRE_THROWS_MATCHES_IMPL(MAYBE_ABORT, EXPRESSION, EXCEPTION, ...)            \
         do {                                                                                       \
-            auto& SNITCH_CURRENT_TEST = snitch::impl::get_current_test();                          \
+            auto& SNITCH_CURRENT_TEST        = snitch::impl::get_current_test();                   \
+            bool  SNITCH_NO_EXCEPTION_THROWN = false;                                              \
             try {                                                                                  \
                 static_cast<void>(EXPRESSION);                                                     \
-                SNITCH_CURRENT_TEST.reg.report_assertion(                                          \
-                    false, SNITCH_CURRENT_TEST, SNITCH_CURRENT_LOCATION,                           \
-                    #EXCEPTION " expected but no exception thrown");                               \
-                MAYBE_ABORT;                                                                       \
+                SNITCH_NO_EXCEPTION_THROWN = true;                                                 \
             } catch (const EXCEPTION& e) {                                                         \
                 auto&& SNITCH_TEMP_MATCHER = __VA_ARGS__;                                          \
                 if (!SNITCH_TEMP_MATCHER.match(e)) {                                               \
@@ -84,6 +86,12 @@
                         #EXCEPTION " expected but other unknown exception thrown");                \
                     MAYBE_ABORT;                                                                   \
                 }                                                                                  \
+            }                                                                                      \
+            if (SNITCH_NO_EXCEPTION_THROWN) {                                                      \
+                SNITCH_CURRENT_TEST.reg.report_assertion(                                          \
+                    false, SNITCH_CURRENT_TEST, SNITCH_CURRENT_LOCATION,                           \
+                    #EXCEPTION " expected but no exception thrown");                               \
+                MAYBE_ABORT;                                                                       \
             }                                                                                      \
         } while (0)
 

--- a/tests/runtime_tests/check.cpp
+++ b/tests/runtime_tests/check.cpp
@@ -1493,6 +1493,21 @@ TEST_CASE("check throws as", "[test macros]") {
         CHECK_EXPR_SUCCESS(catcher);
     }
 
+    SECTION("fail no exception") {
+        std::size_t failure_line = 0u;
+
+        {
+            test_override override(catcher);
+            auto          fun = []() {};
+            // clang-format off
+            SNITCH_CHECK_THROWS_AS(fun(), my_exception); failure_line = __LINE__;
+            // clang-format on
+        }
+
+        CHECK_EXPR_FAILURE(
+            catcher, failure_line, "my_exception expected but no exception thrown"sv);
+    }
+
     SECTION("fail other std::exception") {
         std::size_t failure_line = 0u;
 
@@ -1525,6 +1540,29 @@ TEST_CASE("check throws as", "[test macros]") {
     }
 }
 
+TEST_CASE("require throws as", "[test macros]") {
+    event_catcher<1> catcher;
+
+    SECTION("fail no exception") {
+        std::size_t failure_line = 0u;
+
+        {
+            test_override override(catcher);
+            auto          fun = []() {};
+            try {
+                // clang-format off
+                failure_line = __LINE__; SNITCH_REQUIRE_THROWS_AS(fun(), my_exception);
+                // clang-format on
+            } catch (const snitch::impl::abort_exception&) {
+                // Ignore the abort signal.
+            }
+        }
+
+        CHECK_EXPR_FAILURE(
+            catcher, failure_line, "my_exception expected but no exception thrown"sv);
+    }
+}
+
 TEST_CASE("check throws matches", "[test macros]") {
     event_catcher<1> catcher;
 
@@ -1537,6 +1575,22 @@ TEST_CASE("check throws matches", "[test macros]") {
         }
 
         CHECK_EXPR_SUCCESS(catcher);
+    }
+
+    SECTION("fail no exception") {
+        std::size_t failure_line = 0u;
+
+        {
+            test_override override(catcher);
+            auto          fun     = []() {};
+            auto          matcher = snitch::matchers::with_what_contains{"exception"};
+            // clang-format off
+            failure_line = __LINE__; SNITCH_CHECK_THROWS_MATCHES(fun(), my_exception, matcher);
+            // clang-format on
+        }
+
+        CHECK_EXPR_FAILURE(
+            catcher, failure_line, "my_exception expected but no exception thrown"sv);
     }
 
     SECTION("fail other std::exception") {
@@ -1587,6 +1641,30 @@ TEST_CASE("check throws matches", "[test macros]") {
         CHECK_EXPR_FAILURE(
             catcher, failure_line,
             "could not match caught std::exception with expected content: could not find 'exception1' in 'exception2'"sv);
+    }
+}
+
+TEST_CASE("require throws matches", "[test macros]") {
+    event_catcher<1> catcher;
+
+    SECTION("fail no exception") {
+        std::size_t failure_line = 0u;
+
+        {
+            test_override override(catcher);
+            auto          fun     = []() {};
+            auto          matcher = snitch::matchers::with_what_contains{"exception"};
+            try {
+                // clang-format off
+                failure_line = __LINE__; SNITCH_REQUIRE_THROWS_MATCHES(fun(), my_exception, matcher);
+                // clang-format on
+            } catch (const snitch::impl::abort_exception&) {
+                // Ignore the abort signal.
+            }
+        }
+
+        CHECK_EXPR_FAILURE(
+            catcher, failure_line, "my_exception expected but no exception thrown"sv);
     }
 }
 

--- a/tests/runtime_tests/check.cpp
+++ b/tests/runtime_tests/check.cpp
@@ -530,9 +530,17 @@ SNITCH_WARNING_PUSH
 SNITCH_WARNING_DISABLE_PRECEDENCE
 SNITCH_WARNING_DISABLE_ASSIGNMENT
 
+#if defined(SNITCH_COMPILER_MSVC) && _MSC_VER == 1937
+// Regression in MSVC compiler 19.37.*
+// https://github.com/snitch-org/snitch/issues/140
+// https://developercommunity.visualstudio.com/t/Regression:-False-positive-C7595:-std::/10509214
+#    define SNITCH_TEST_NO_SPACESHIP
+#endif
+
 TEST_CASE("check no decomposition", "[test macros]") {
     event_catcher<1> catcher;
 
+#if !defined(SNITCH_TEST_NO_SPACESHIP)
     SECTION("spaceship") {
         int         value1       = 1;
         int         value2       = 1;
@@ -549,6 +557,7 @@ TEST_CASE("check no decomposition", "[test macros]") {
         CHECK(value2 == 1);
         CHECK_EXPR_FAILURE(catcher, failure_line, "CHECK"sv, "value1 <=> value2 != 0"sv, ""sv);
     }
+#endif
 
     SECTION("with operator &&") {
         int         value1       = 1;

